include/arch/riscv/thread.h

```diff
@@ -30,6 +30,46 @@
 #endif
 #endif
 
+#ifdef CONFIG_RISCV_PMP
+#ifdef CONFIG_64BIT
+#define	RISCV_PMP_CFG_NUM	(CONFIG_PMP_SLOT >> 3)
+#else
+#define	RISCV_PMP_CFG_NUM	(CONFIG_PMP_SLOT >> 2)
+#endif
+#endif
+
+#ifdef CONFIG_RISCV_PMP
+/*
+ * PMP entries:
+ *   (1 for interrupt stack guard: None)
+ *   4 for stacks guard: None
+ *   1 for RAM: RW
+ *   1 for other address space: RWX
+ */
+#define PMP_REGION_NUM_FOR_STACK_GUARD	6
+#define PMP_CFG_CSR_NUM_FOR_STACK_GUARD	2
+#endif
+
+#ifdef CONFIG_USERSPACE
+#ifdef CONFIG_PMP_STACK_GUARD
+/*
+ * 1 for interrupt stack guard: None
+ * 1 for core state: R
+ * 2 for program and read only data: RX
+ * 2 for user thread stack: RW
+ */
+#define PMP_REGION_NUM_FOR_U_THREAD	6
+#else
+/*
+ * 1 for core state: R
+ * 2 for program and read only data: RX
+ * 2 for user thread stack: RW
+ */
+#define PMP_REGION_NUM_FOR_U_THREAD	5
+#endif
+#define PMP_MAX_DYNAMIC_REGION	(CONFIG_PMP_SLOT - PMP_REGION_NUM_FOR_U_THREAD)
```

Comment:
Question about this item.

The default CONFIG_PMP_SLOT is 8, so this by default is 3.

```
int arch_mem_domain_max_partitions_get(void)
{
	return PMP_MAX_DYNAMIC_REGION;
}
```

This means that the kernel thinks that we can have up to 3 k_mem_partitions programmed in a memory domain.

But this code sure looks like any given memory partition requires 2 PMP slots. Is just one needed per k_mem_partition? If so, why does the user thread stack (for example) require two slots?

Reply:
> But this code sure looks like any given memory partition requires 2 PMP slots
No, it can use 1 PMP slot if size is equal 4 or power of 2.
https://github.com/zephyrproject-rtos/zephyr/pull/27063/files#diff-31937c1b5ccaf40e621e91589a7b51c0R396

Reply:
If that is the case why are you not setting CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT and CONFIG_GEN_PRIV_STACKS?

You should only need 1 region for the thread stack as well since with the above configs enabled the thread stacks will also be power of two sized/aligned

Without these settings, automatic memory partitions will NOT be sized correctly

Reply:
https://docs.zephyrproject.org/latest/guides/porting/arch.html#stack-objects

Reply:
in linker.ld, you have:

```
#if defined(CONFIG_USERSPACE)
#define APP_SHARED_ALIGN . = ALIGN(4)
#define SMEM_PARTITION_ALIGN(region_size) . = ALIGN(4)

#include <app_smem.ld>
```

Which means that any given automatic memory partition is going to be 4-byte aligned and not to some power of two.
Which means two PMP slots per partition.
Need you to carefully check all of this.

Reply:
To sum up: if you want to give more flexibility on stack object sizes and use two PMP for the stack, that's up to you, although it would make another region available if you didn't.

Consider the same for text/rodata region since IIRC you have XIP and this is on flash. Would make another slot available if you just used one.

But the memory domain partitions on the other hand have to be set up correctly. Either the code has to enforce these requirements for the automatic memory partitions, or you have two use two slots for each one. The latter would make this unusable in practice.

I'm not sure how this passed tests in this state.

Reply:
> This option is enabled when the MPU requires a power of two alignment

https://github.com/zephyrproject-rtos/zephyr/blob/5e0b55c30e52fa9341f4c91f1bd083ef82a8b6ec/arch/Kconfig#L558
 Is it not a requirement, just a slot optimization on running time. I will take a look deeper later (monday) but if the framework must have fixed value.
For thread value slot, we can at least have 4 slot if we force power of two.

Reply:
Naturally if I've missed something please do point it out.

Reply:
I think you are trying to optimize stuff too early. Let's get everything *correct* first. Then look at opportunities for optimization, which may include enhancements to the core kernel to support the 1 or 2 slot capability of this arch (which could also be used to allow for more flexibility on partition sizes on platforms like ARMv7-M)

Reply:
>> just a slot optimization on running time

> I don't understand what you mean by this. Can you please clarify?

If the address start and size of a memory domain are compatible with power of 2, then I setup a NAPOT instead of NA4 + TOR to save one PMP slot and increase a bit the performance for the boards like hifive1_revb (probably all CORE_E31). This piece of code shouldn't have an impact on core kernel memory domain implementation.

Ok I think I've understood your point (hoping), there is indeed a problem. I prefer not force power of two alignment because today 8 slot is enough to have a working userspace + 1 memory domain without NAPOT, and that save memory space which is really small on current supported board (and QEMU). So, to fix that, I suggest to:
- Divide per two the current PMP_MAX_DYNAMIC_REGION formula (This is for me the issue).
- Remove the memory domain NAPOT usage if you think that impact the core kernel memory domain implementation (but I'm sure it doesn't).

In case of someone need more shared memory:
- Firstly, I think is it cheaper to add PMP slot rather than add more memory space. Then it is more likely that the next generation of core have 16 slots.
- Secondly, it still possible to use NAPOT, just need to patch the current code. In fact, a (huge) improvement could be add the power of two alignment support, which could be enable through RISCV Kconfig flag.

Do you agree ?

Reply:
Adding MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT w/ linker script change increases the number of partitions supported to 2. Enforcing power of 2 stack increases it further to 3. These seem like simple and worthwhile changes.

Reply:
@xodus Yes but you forgot extra memory usage (I already had lack of memory issue during test). I think both (with and without power of two are needed) are worth (depend where is your limitation). Here I provide without power of two which is working. Lets do improvement in a second time "even if it's simple".

Reply:
@amergnat the merge window for 2.4 is now closed. This won't be considered for merge until after it opens again at the end of the month. You might as well make these changes now so that this implementation is usable on the current generation of CPUs.

> Lets do improvement in a second time "even if it's simple".

I don't agree with this approach given the circumstances.

Reply:
Then I will implement it no problem

> You might as well make these changes now so that this implementation is usable on the current generation of CPUs.

I probably misspoke, then you misunderstood. My current implementation is usable on the current generation of CPUs, just saying it should be better than forcing power of two on the next generation.

As I said, I will try to implement it with a Kconfig flag. That allow user to chose power of two or not.

Reply:
you're aware that only allowing 1 partition in a memory domain results in something unusable in practice right?

Reply:
No I wasn't. I was thinking that could be ok for some usage. Thanks.

Reply:
Fixed